auth-api/src/auth_api/services/products.py

```diff
@@ -259,3 +263,90 @@ def send_approved_product_subscription_notification(receipt_admin_emails, produc
+                                 f'Keycloak Guid: {kgs.user_guid}')
+        KeycloakService.add_or_remove_user_from_group(kgs.user_guid, kgs.group_name, kgs.group_action)
+        current_app.logger.debug('>add_and_remove_user_to_product_keycloak_roles ')
+
+    @staticmethod
+    def get_users_product_subscriptions_kc_groups(user_ids: List[int]) -> List[KeycloakGroupSubscription]:
+        """Generate Keycloak Group Subscriptions."""
+        ps_max_subquery = db.session.query(
+            func.max(ProductSubscriptionModel.id).label('id'),
+            ProductSubscriptionModel.product_code,
+            ProductSubscriptionModel.org_id
+        ) \
+            .group_by(ProductSubscriptionModel.product_code, ProductSubscriptionModel.org_id) \
+            .subquery()
+
+        m_max_subquery = db.session.query(
+            func.max(MembershipModel.id).label('id'),
+            MembershipModel.org_id,
+            MembershipModel.user_id
+        ) \
+            .group_by(MembershipModel.org_id, MembershipModel.user_id) \
+            .subquery()
+
+        active_subscription_case = case(
+            [
+                 (and_(MembershipModel.status == Status.ACTIVE.value, ProductSubscriptionModel.status_code ==
+                       ProductSubscriptionStatus.ACTIVE.value), 1),
+            ],
+            else_=0
+        )
+
+        user_subscriptions = db.session.query(UserModel, ProductCodeModel) \
+            .join(ProductCodeModel, literal(True)) \
+            .outerjoin(m_max_subquery, m_max_subquery.c.user_id == UserModel.id) \
+            .outerjoin(MembershipModel, MembershipModel.id == m_max_subquery.c.id) \
+            .outerjoin(
+            # pylint: disable=comparison-with-callable
+            ps_max_subquery, ps_max_subquery.c.product_code == ProductCodeModel.code) \
+            .outerjoin(ProductSubscriptionModel, ProductSubscriptionModel.id == ps_max_subquery.c.id) \
+            .filter(or_(
+                    ProductSubscriptionModel.org_id == MembershipModel.org_id,
+                    ProductSubscriptionModel.org_id.is_(None),
+                    MembershipModel.org_id.is_(None)))\
+            .filter(UserModel.id.in_(user_ids)) \
+            .filter(ProductCodeModel.keycloak_group.isnot(None))\
+            .group_by(UserModel.id, UserModel.keycloak_guid, ProductCodeModel.code, ProductCodeModel.keycloak_group)\
+            .order_by(UserModel.id, ProductCodeModel.code)\
+            .with_entities(
+                UserModel.id,
+                UserModel.keycloak_guid,
+                ProductCodeModel.code,
+                ProductCodeModel.keycloak_group,
+                func.sum(active_subscription_case).label('active_subscription_count')
+        ).all()  # pylint: disable=comparison-with-callable
+
+        keycloak_group_subscriptions = []
+        for ups in user_subscriptions:
+            action = KeycloakGroupActions.ADD_TO_GROUP.value \
```

Comment:
There might have been a way to create a case to make this unnecessary, but I'm not entirely sure how to do nested cases in SQLAlchemy